osu.Game/Screens/Play/PlayerTouchInputHandler.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Game.Configuration;
+using osu.Game.Overlays;
+using osu.Game.Overlays.OSD;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Screens.Play
+{
+    public partial class PlayerTouchInputHandler : Component
+    {
+        [Resolved]
+        private Player player { get; set; } = null!;
+
+        [Resolved]
+        private GameplayState gameplayState { get; set; } = null!;
+
+        [Resolved]
+        private OnScreenDisplay? onScreenDisplay { get; set; }
+
+        private IBindable<bool> touchActive = new BindableBool();
+
+        [BackgroundDependencyLoader]
+        private void load(SessionStatics statics)
+        {
+            touchActive = statics.GetBindable<bool>(Static.TouchInputActive);
+            touchActive.BindValueChanged(_ => updateState());
+        }
+
+        private void updateState()
+        {
+            if (!touchActive.Value)
+                return;
+
+            if (gameplayState.HasPassed || gameplayState.HasFailed || gameplayState.HasQuit)
+                return;
+
+            if (gameplayState.Score.ScoreInfo.Mods.OfType<ModTouchDevice>().Any())
+                return;
+
+            if (player.IsBreakTime.Value)
+                return;
+
+            var touchDeviceMod = gameplayState.Ruleset.GetTouchDeviceMod();
+            if (touchDeviceMod == null)
+                return;
+
+            // do not show the toast if the user hasn't hit anything yet.
+            // we're kind of assuming that the user just switches to touch for gameplay
+            // and we don't want to spam them with obvious toasts.
+            if (gameplayState.ScoreProcessor.HitEvents.Any(ev => ev.Result.IsHit()))
```

Comment:
I'm not sure about the precondition here... it means there's a chance the user can turn on TD but it doesn't display, but I don't really understand why we'd want that.

In addition, I'm not sure about showing this at all, it might be considered a distraction to users. As far as I can tell, this will display every play for a user that uses mouse at song select and touch only during gameplay, yeah?

Reply:
> In addition, I'm not sure about showing this at all, it might be considered a distraction to users.

Hmm. I only bothered because stable did this, but I failed to notice that it only [does so in debug](https://github.com/peppy/osu-stable-reference/blob/3ea48705eb67172c430371dcfc8a16a002ed0d3d/osu!/GameModes/Play/Player.cs#L4102-L4104).

Can remove entirely to get rid of this conundrum entirely.

> it means there's a chance the user can turn on TD but it doesn't display, but I don't really understand why we'd want that.

My primary goal was to inform a user when they "downgrade" their submittable play from normal to touch-device-laden. So currently it will not show if the user uses mouse at song select but starts hitting objects with touch.

Reply:
I think not showing at all is likely fine for now. The most common case will see the user aware of touch mod at song select / loading anyways.

Reply:
Snipped in https://github.com/ppy/osu/pull/25348/commits/682668ccf0f73bdbb98b800afcaff983399d391d.